tasm/tiles/src/LayoutDatabase.cc

```diff
@@ -0,0 +1,386 @@
+    }
+}
+
+void LayoutDatabase::createHeightsTable() {
+    const char *createHeightTable = "CREATE TABLE heights ("\
+                            "video TEXT NOT NULL,"\
+                            "id INT NOT NULL,"\
+                            "height INT NOT NULL,"\
+                            "position INT NOT NULL,"\
+                            "FOREIGN KEY (video, id) REFERENCES layouts(video, id))";
+    char *error = nullptr;
+    int result = sqlite3_exec(db_, createHeightTable, NULL, NULL, &error);
+    if (result != SQLITE_OK) {
+        std::cerr << "Error creating layout height table: " << sqlite3_errmsg(db_) << std::endl;;
+        sqlite3_free(error);
+    }
+
+    const char *createIndex = "CREATE INDEX heights_index ON heights(video, id)";
+    result = sqlite3_exec(db_, createIndex, NULL, NULL, &error);
+    if (result != SQLITE_OK) {
+        std::cerr << "Error creating heights index: " << sqlite3_errmsg(db_) << std::endl;
+        sqlite3_free(error);
+    }
+}
+
+std::vector<int> LayoutDatabase::tileLayoutIdsForFrame(const std::string &video, unsigned int frameNumber) const {
+    std::string selectLayoutIds = "SELECT id FROM layouts WHERE video = ? AND ? BETWEEN first_frame AND last_frame";
+
+    sqlite3_stmt *select;
+    ASSERT_SQLITE_OK(sqlite3_prepare_v2(db_, selectLayoutIds.c_str(), selectLayoutIds.length(), &select, nullptr));
+    ASSERT_SQLITE_OK(sqlite3_bind_text(select, 1, video.c_str(), -1, SQLITE_STATIC));
+    ASSERT_SQLITE_OK(sqlite3_bind_int(select, 2, frameNumber));
+
+    int result;
+    std::vector<int> layoutIds;
+    while ((result = sqlite3_step(select)) == SQLITE_ROW) {
+        layoutIds.push_back(sqlite3_column_int(select, 0));
+    }
+
+    ASSERT_SQLITE_DONE(result);
+    ASSERT_SQLITE_OK(sqlite3_finalize(select));
+
+    assert(!layoutIds.empty());
+    return layoutIds;
+}
+
+std::shared_ptr<TileLayout> LayoutDatabase::tileLayoutForId(const std::string &video, int id) const {
+    // select number of rows and cols
+    std::string query = "SELECT num_rows, num_cols FROM layouts WHERE video = ? AND id = ?";
+    sqlite3_stmt *select;
+    ASSERT_SQLITE_OK(sqlite3_prepare_v2(db_, query.c_str(), query.length(), &select, nullptr));
+    ASSERT_SQLITE_OK(sqlite3_bind_text(select, 1, video.c_str(), -1, SQLITE_STATIC));
+    ASSERT_SQLITE_OK(sqlite3_bind_int(select, 2, id));
+
+    int result;
+    unsigned int numberOfRows = 0;
+    unsigned int numberOfCols = 0;
+    if ((result = sqlite3_step(select)) == SQLITE_ROW) {
+        numberOfRows = sqlite3_column_int(select, 0);
+        numberOfCols = sqlite3_column_int(select, 1);
+    }
+
+    ASSERT_SQLITE_DONE(sqlite3_step(select));
+    ASSERT_SQLITE_OK(sqlite3_finalize(select));
+
+    std::vector<unsigned int> heightsOfRows = heightsForId(video, id);
+
+    std::vector<unsigned int> widthsOfColumns = widthsForId(video, id);
+
+    return std::make_shared<TileLayout>(TileLayout(numberOfCols, numberOfRows, widthsOfColumns, heightsOfRows));
```

Comment:
Because we don't use heightsOfRows and widthsOfColumns, you could inline their calls. Also, similar to before I think the compiler may be able to figure out how to construct the TileLayout just given:
`std::make_shared<TileLayout>(numberOfCols, numberOfRows, widthsForId(video, id), heightsForId(video, id));`